### PR TITLE
fix(governance): Use error boundaries when creating governance tx

### PR DESCRIPTION
**Pull request by michalszorad**
Resolves [YOMO-1155](https://emurgo.atlassian.net/browse/YOMO-1155)

[YOMO-1155]: https://emurgo.atlassian.net/browse/YOMO-1155?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jorbuedo — APPROVED** [review]

File: apps/wallet-mobile/src/features/Staking/Governance/common/helpers.ts [hunk 1]
```diff
@@ -20,6 +22,15 @@ export const useIsParticipatingInGovernance = (wallet: YoroiWallet) => {
...
 
+export const useCreateGovernanceTx = (wallet: YoroiWallet) => {
+  const mutationFn = (certificates: Certificate[]) => {
+    return wallet.createUnsignedGovernanceTx(certificates)
+  }
+
+  const mutation = useMutation({mutationFn, useErrorBoundary: true, retry: false})
+  return {...mutation, createUnsignedGovernanceTx: mutation.mutateAsync}
+}
+
```
Comment (banklesss) [inline apps/wallet-mobile/src/features/Staking/Governance/common/helpers.ts]: Would it be interesting to have this hook in the hooks file?

Comment (banklesss) [reply apps/wallet-mobile/src/features/Staking/Governance/common/helpers.ts]: I see the hooks there as the bridge between react and cardano in general 🤔

Comment (michalszorad) [reply apps/wallet-mobile/src/features/Staking/Governance/common/helpers.ts]: Hello, do you mean in the `@yoroi/staking` package?

Comment (banklesss) [reply apps/wallet-mobile/src/features/Staking/Governance/common/helpers.ts]: here -> `src/yoroi-wallets/hooks`

Comment (michalszorad) [reply apps/wallet-mobile/src/features/Staking/Governance/common/helpers.ts]: Yeah, sounds good. Will update

**Review by banklesss — APPROVED** [review]

File: apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts [hunk 2]
```diff
@@ -1090,3 +1091,12 @@ export const useNativeAssetInvalidation = ({networkId, policy, name}: NativeAsse
...
   }
 }
+
+export const useCreateGovernanceTx = (wallet: YoroiWallet) => {
+  const mutationFn = (certificates: Certificate[]) => {
+    return wallet.createUnsignedGovernanceTx(certificates)
+  }
+
+  const mutation = useMutation({mutationFn, useErrorBoundary: true, retry: false})
+  return {...mutation, createUnsignedGovernanceTx: mutation.mutateAsync}
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts]: the aliases should be on mutationFn, the mutateAsync is wrapped by query, it should consume from mutation.mutateAsync on the client only, yet, be aware that none of the status flag will reflect -> isLoading, isError, while using it.

File: apps/wallet-mobile/src/features/Staking/Governance/useCases/Home/HomeScreen.tsx [hunk 3]
```diff
@@ -266,13 +266,8 @@ const NeverParticipatedInGovernanceVariant = () => {
...
             : null
           const certs = stakeCert !== null ? [stakeCert, certificate] : [certificate]
-          const unsignedTx = await wallet.createUnsignedGovernanceTx(certs)
-          navigateTo.confirmTx({
-            unsignedTx,
-            vote: {kind: 'no-confidence'},
-            registerStakingKey: stakeCert !== null,
-            navigateToStakingOnSuccess,
-          })
+          const unsignedTx = await createUnsignedGovernanceTx(certs)
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Staking/Governance/useCases/Home/HomeScreen.tsx]: get the async version here from the mutation.

File: apps/wallet-mobile/src/features/Staking/Governance/useCases/Home/HomeScreen.tsx [hunk 4]
```diff
@@ -221,13 +229,8 @@ const NeverParticipatedInGovernanceVariant = () => {
...
               : null
             const certs = stakeCert !== null ? [stakeCert, certificate] : [certificate]
-            const unsignedTx = await wallet.createUnsignedGovernanceTx(certs)
-            navigateTo.confirmTx({
-              unsignedTx,
-              vote: {kind: 'delegate', drepID},
-              registerStakingKey: stakeCert !== null,
-              navigateToStakingOnSuccess,
-            })
+            const unsignedTx = await createUnsignedGovernanceTx(certs)
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Staking/Governance/useCases/Home/HomeScreen.tsx]: and here

File: apps/wallet-mobile/src/features/Staking/Governance/useCases/ChangeVote/ChangeVoteScreen.tsx [hunk 5]
```diff
@@ -63,25 +66,27 @@ export const ChangeVoteScreen = () => {
...
   const handleAbstain = async () => {
     const stakingKey = await wallet.getStakingKey()
+    const vote = {kind: 'abstain'} as const
     createVotingCertificate(
       {vote: 'abstain', stakingKey},
       {
         onSuccess: async (certificate) => {
-          const unsignedTx = await wallet.createUnsignedGovernanceTx([certificate])
-          navigateTo.confirmTx({unsignedTx, vote: {kind: 'abstain'}})
+          const unsignedTx = await createUnsignedGovernanceTx([certificate])
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Staking/Governance/useCases/ChangeVote/ChangeVoteScreen.tsx]: ◀️

File: apps/wallet-mobile/src/features/Staking/Governance/useCases/ChangeVote/ChangeVoteScreen.tsx [hunk 6]
```diff
@@ -63,25 +66,27 @@ export const ChangeVoteScreen = () => {
...
   const handleNoConfidence = async () => {
     const stakingKey = await wallet.getStakingKey()
+    const vote = {kind: 'no-confidence'} as const
     createVotingCertificate(
       {vote: 'no-confidence', stakingKey},
       {
         onSuccess: async (certificate) => {
-          const unsignedTx = await wallet.createUnsignedGovernanceTx([certificate])
-          navigateTo.confirmTx({unsignedTx, vote: {kind: 'no-confidence'}})
+          const unsignedTx = await createUnsignedGovernanceTx([certificate])
```
Comment (stackchain) [inline apps/wallet-mobile/src/features/Staking/Governance/useCases/ChangeVote/ChangeVoteScreen.tsx]: ◀️

File: apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts (hunk 2 above)
Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts]: Having a look

Comment (michalszorad) [reply apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts]: That's now updated everywhere

**Review by stackchain — APPROVED** [review]

**Pull request merged**